### PR TITLE
PT-1840: Validate child properties for catalog and categories

### PR DESCRIPTION
**Pull request by pushnitsa**
### Jira-link:

https://virtocommerce.atlassian.net/browse/PT-1840

### CODE REVIEW

File: src/VirtoCommerce.CatalogModule.Data/Validation/CategoryPropertyNameValidator.cs [hunk 1]
```diff
@@ -1,36 +1,122 @@
...
-            {
-                var propertySearchCriteria = new PropertySearchCriteria
+            RuleFor(r => r)
+                .CustomAsync(async (r, context, _) =>
                 {
-                    PropertyNames = new[] { r.PropertyName },
-                    CategoryId = r.CategoryId
-                };
-                var propertySearchResult = await _propertySearchService.SearchPropertiesAsync(propertySearchCriteria);
+                    (var isPropertyUniqueForHierarchy, var CategoryName) = await CheckPropertyUniquenessAsync(r);
```
Comment (Vectorfield4) [inline src/VirtoCommerce.CatalogModule.Data/Validation/CategoryPropertyNameValidator.cs:48]:
```suggestion
                    (var isPropertyUniqueForHierarchy, var categoryName) = await CheckPropertyUniquenessAsync(r);
```

File: src/VirtoCommerce.CatalogModule.Data/Validation/CategoryPropertyNameValidator.cs [hunk 2]
```diff
@@ -1,36 +1,122 @@
...
+                        });
+                    }
+                });
+        }
+
+        protected virtual async Task<(bool, string)> CheckPropertyUniquenessAsync(CategoryPropertyValidationRequest request)
+        {
+            var useIndexedSearch = await _settingsManager.GetValueAsync(catalogCore.ModuleConstants.Settings.Search.UseCatalogIndexedSearchInManager.Name, true);
+
+            var categoryIds = new List<string>();
```
Comment (Vectorfield4) [inline src/VirtoCommerce.CatalogModule.Data/Validation/CategoryPropertyNameValidator.cs:68]: Can we move this block to the protected method?
image

**Review by Vectorfield4 — APPROVED** [review]

File: src/VirtoCommerce.CatalogModule.Core/Model/CategoryPropertyValidationRequest.cs [hunk 3]
```diff
@@ -1,13 +1,11 @@
-using System;
-using System.Collections.Generic;
-using System.Text;
-
 namespace VirtoCommerce.CatalogModule.Core.Model
 {
     public class CategoryPropertyValidationRequest
     {
```
Comment (ksavosteev) [inline src/VirtoCommerce.CatalogModule.Core/Model/CategoryPropertyValidationRequest.cs:5]: remove empty line

**Review by ksavosteev — APPROVED** [review]

File: tests/VirtoCommerce.CatalogModule.Tests/CategoryIteratorTests.cs [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+using VirtoCommerce.CatalogModule.Data.Repositories;
+using VirtoCommerce.CatalogModule.Data.Services;
+using VirtoCommerce.Platform.Core.Settings;
+using Xunit;
+
+namespace VirtoCommerce.CatalogModule.Tests
+{
+    public class CategoryIteratorTests
+    {
+        private readonly Fixture _fixture = new();
```
Comment (Vectorfield4) [inline tests/VirtoCommerce.CatalogModule.Tests/CategoryIteratorTests.cs:19]: Unavailable feature in .net core 3.1. 
```suggestion
        private readonly Fixture _fixture = new Fixture();
```

**Review by Vectorfield4 — APPROVED** [review]

File: src/VirtoCommerce.CatalogModule.Data/Services/CategoryIterator.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+            _handledItemsCount += result.Count();
+
+            HasMoreResults = _handledItemsCount < _totalCount;
+
+            return result.ToImmutableArray();
+        }
+
+
+        private async Task EnsureThatDataIsActualAsync()
```
Comment (ksavosteev) [inline src/VirtoCommerce.CatalogModule.Data/Services/CategoryIterator.cs:69]: rename to LoadData

File: src/VirtoCommerce.CatalogModule.Data/Services/CategoryIterator.cs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+using VirtoCommerce.CatalogModule.Core;
+using VirtoCommerce.CatalogModule.Core.Model.Search;
+using VirtoCommerce.CatalogModule.Core.Search;
+using VirtoCommerce.CatalogModule.Data.Repositories;
+using VirtoCommerce.Platform.Core.Common;
+using VirtoCommerce.Platform.Core.Settings;
+
+namespace VirtoCommerce.CatalogModule.Data.Services
+{
+    public sealed class CategoryIterator
```
Comment (ksavosteev) [inline src/VirtoCommerce.CatalogModule.Data/Services/CategoryIterator.cs:15]: rename to CategoryHierarchyIterator

File: src/VirtoCommerce.CatalogModule.Data/Services/CategoryIteratorFactory.cs [hunk 7]
```diff
@@ -0,0 +1,33 @@
+using System;
+using VirtoCommerce.CatalogModule.Core.Search;
+using VirtoCommerce.CatalogModule.Data.Repositories;
+using VirtoCommerce.Platform.Core.Settings;
+
+namespace VirtoCommerce.CatalogModule.Data.Services
+{
+    public sealed class CategoryIteratorFactory
```
Comment (ksavosteev) [inline src/VirtoCommerce.CatalogModule.Data/Services/CategoryIteratorFactory.cs:8]: rename to CategoryHierarchyIteratorFactory
or consider removing this class and using Func<> factory registration in ServiceCollection

**Review by ksavosteev — APPROVED** [review]

**Pull request merged**